doc/Architecture.md

```diff
@@ -61,158 +61,210 @@ TODO/FIXME: Everything here assumes that domain ownership is eternal/indefinite.
+
+- One may consider to not allow `DeNSKey`s with empty strings to be inserted.
+
+#### SetValidator validator
+`SetValidator` is the validator address which holds the elements in the logical database. It simply validates only if `SetElemID` mints i.e., it forwards all verifications to `SetElemID`.
+
+**DISCUSSION:**
+Since `SetValidator` forwards everything to `SetElemID`, this makes the potential for upgrading to a new `SetElemID` convenient as swapping to a new `SetElemID'` would still keep existing `SetElemID` already at `SetValidator` meaning there would be no need for a large migration of tokens.
+
+#### ElementID Minting Policy
+
+The `ElementID` minting policy is an NFT which serves as an access control identifier in order to gate access to specific records in the virtual DeNS database.
+
+**CHECKS:**:
+
+- `SetElemID` mints in the same transaction
+
+**DISCUSSION:** 
+
+This forwards all verifications to `SetElemID` whose minting is unique w.r.t to the element inserted in the set, and ensures that this `ElementID`'s token name contains sufficient information to identify it with the inserted element.
 
 ### Records Validator
 
-The records validator serves as an access control list for the virtual DeNS database. When the offchain code reconstructs the database from Cardano keys and Arweave values, it treats the records locked at the Records validator as the source of authority. That is: Only records locked at the records validator are used to construct the database used in a DeNS resolver.
+The records validator serves as an access control list for the virtual DeNS database. When the offchain code reconstructs the database from Cardano keys and Arweave values, it treats the records locked at the Records validator identified with a `ElementID` token as the source of authority. That is: Only records locked at the records validator are used to construct the database used in a DeNS resolver.
 
 **PARAMETERS**:
-  - The Protocol NFT
-  - The CurrencySymbol/ScriptHash of the ElementID minting policy
+
+- The Protocol NFT
 
 **DATUM**:
 
-The the Records Validator datum functions as the key in our virtual database. To update or create a record, a DeNS user simply adjusts the reference for the class and name that they own. To delete a record, the user submits a transaction with an output datum where the `reference` field is `Nothing`.
+The Records Validator datum functions as the key in our virtual database. To update or create a record, a DeNS user simply adjusts the reference for the class and name that they own. To delete a record, the user submits a transaction with an output datum where the `reference` field is `Nothing`.
 
-``` haskell
-data DeNSKey = DeNSKey {
-  class :: Word32,
-  name  :: ByteString,
-  reference :: Maybe ByteString -- an Arweave address
+```haskell
+data DeNSKey = DeNSKey 
+    { class :: Word32
+    , name  :: ByteString
+    , reference :: Maybe ByteString 
+        -- ^ an Arweave address
+    , owner :: PubKeyHash
```

Comment:
Just want to emphasize:
- The old version let anyone update. I really don't think we want this. Instead, this proposes to let the owner (some public key hash) verify that the update is valid.